src/proxy/resolve.rs

```diff
@@ -107,61 +125,440 @@ where
 
     fn call(&mut self, target: T) -> Self::Future {
         let resolution = self.resolve.resolve(&target);
-        futures::future::ok(Discover {
+        futures::future::ok(Discover::new(resolution, self.inner.clone()))
+    }
+}
+
+// === impl Discover ===
+
+impl<R, M> Discover<R, M>
+where
+    R: Resolution,
+    M: svc::Service<R::Endpoint>,
+{
+    fn new(resolution: R, make: M) -> Self {
+        Self {
             resolution,
-            make: self.inner.clone(),
+            make,
+            make_futures: MakeFutures::new(),
             is_empty: Arc::new(AtomicBool::new(false)),
-        })
+        }
     }
 }
 
 impl<R, M> HasEndpointStatus for Discover<R, M>
 where
     R: Resolution,
+    M: svc::Service<R::Endpoint>,
 {
     fn endpoint_status(&self) -> EndpointStatus {
         EndpointStatus(self.is_empty.clone())
     }
 }
 
-impl<R, M> tower_discover::Discover for Discover<R, M>
+impl<R, M> Discover<R, M>
 where
     R: Resolution,
     R::Endpoint: fmt::Debug,
     R::Error: Into<Error>,
-    M: rt::Make<R::Endpoint>,
+    M: svc::Service<R::Endpoint>,
+    M::Error: Into<Error>,
 {
-    type Key = SocketAddr;
-    type Service = M::Value;
-    type Error = Error;
-
-    fn poll(&mut self) -> Poll<Change<Self::Key, Self::Service>, Self::Error> {
+    fn poll_resolution(&mut self) -> Poll<Change<SocketAddr, M::Response>, Error> {
         loop {
-            let up = try_ready!(self.resolution.poll().map_err(Into::into));
-            trace!("watch: {:?}", up);
-            match up {
+            // Before polling the resolution, where we could potentially receive
+            // an `Add`, poll_ready to ensure that `make` is ready to build new
+            // services. Don't process any updates until we can do so.
+            try_ready!(self.make.poll_ready().map_err(Into::into));
+
+            let update = try_ready!(self.resolution.poll().map_err(Into::into));
+            trace!("watch: {:?}", update);
+            match update {
                 Update::Add(addr, target) => {
-                    // We expect the load balancer to handle duplicate inserts
-                    // by replacing the old endpoint with the new one, so
-                    // insertions of new endpoints and metadata changes for
-                    // existing ones can be handled in the same way.
-                    let svc = self.make.make(&target);
-                    self.is_empty.store(false, Ordering::Release);
-                    return Ok(Async::Ready(Change::Insert(addr, svc)));
+                    // Start building the service and continue.
+                    // Remove any pending additions for this addr so that they
+                    // cannot race.
+                    self.make_futures.remove(&addr);
```

Comment:
TIOLI: Seems like this should just be part of `MakeFutures::push` to me --- the main reason I made it a separate struct rather than a new field on `Discover` was that it seemed neater to bundle all the logic for managing the pending service futures together...